### PR TITLE
updated for multiD arrays

### PR DESCRIPTION
**Pull request by vitskvara**
This is the currently working way to do the splitting for multi-D arrays - see https://github.com/aicenter/ConditionalDists.jl/issues/. However there are still some things to be edited, e.g. likelihoods.

### CODE REVIEW

**Comment by vitskvara** [comment]
I have some issues with the shapes of loglikelihood output, we should decide what the behaviour should be, it is not consistent right now.

File: src/cmeanvar_gaussian.jl [hunk 1]
```diff
@@ -53,8 +51,8 @@ function mean_var(p::CMeanVarGaussian{T,ScalarVar}, z::AbstractArray) where T
         error("Mapping should return eltype $T. Found: $(eltype(ex))")
     end
 
-    μ = ex[1:end-1,:]
-    σ = ex[end:end,:] .* fill!(similar(μ, 1, size(μ,2)), 1)
+    μ, σ = last_split(ex)
+    σ = σ # .* fill!(similar(μ, 1, size(μ)[2:end]...), 1) # everything seems to work fine without this...
```
Comment (vitskvara) [inline src/cmeanvar_gaussian.jl:55]: This seems to be unnecessary.

File: test/cmeanvar_gaussian.jl [hunk 2]
```diff
@@ -22,9 +22,23 @@
...
+        z = randn(T, h, w, c, n) |> gpu;
+        μx = mean(p, z);
+        σ2 = variance(p, z);
+        @test mean_var(p, z) == (μx, σ2)
+        @test size(μx) == (h, w, oc, batch)
+        @test size(σ2) == (h, w, 1, batch)
+        @test size(rand(p, z)) == (h, w, oc, batch)
+
+        x = randn(h, w, oc, batch) |> gpu;
+        @test size(loglikelihood(p, x, z)) == (1, 1, 1, batch) # i would like it more to be shape (1, batch)
```
Comment (vitskvara) [inline test/cmeanvar_gaussian.jl:40]: Like stated in the comment, would it be better to have a (1, batch) output size for loglikelihoods of all possible inputs? That would require a dropdims in the loglikelihood definition.

File: src/cmeanvar_gaussian.jl (hunk 1 above)
Comment (nmheim) [reply src/cmeanvar_gaussian.jl:55]: did you test it on gpu? the `fill!` was necessary to make it work there. maybe that changed with the new flux/zygote/cuarrays though...

**Comment by nmheim** [comment]
I am wondering if it would not be easier/more general if we enforce x/z to be vectors.
in the case of e.g. a deconvolutional decoder we would only have to do a couple of reshapes before/after the decoder which are essentially for free and we would not have to deal with all the complications with n-dimensional variances....

@smidl what do you think?

File: src/cmeanvar_gaussian.jl (hunk 1 above)
Comment (vitskvara) [reply src/cmeanvar_gaussian.jl:55]: Yes, this seems to be working even on gpu now.

**Comment by vitskvara** [comment]
> I am wondering if it would not be easier/more general if we enforce x/z to be vectors.
> in the case of e.g. a deconvolutional decoder we would only have to do a couple of reshapes before/after the decoder which are essentially for free and we would not have to deal with all the complications with n-dimensional variances....
> 
> @smidl what do you think?

Yeah, that makes a lot of sense... we will avoid having to deal with similar issues. The responsibility for making things right will be on the user instead.

**Comment by vitskvara** [comment]
I've decided to abandon this for now and just use vectors for everything.

**Pull request closed without merging**